MetaU Capstone Project/API/SpotifyManager.m

```diff
@@ -105,9 +112,147 @@ - (void)appRemote:(nonnull SPTAppRemote *)appRemote didFailConnectionAttemptWith
     NSLog(@"Track name: %@", playerState.track.name);
     NSLog(@"player state changed");
+    
+    [self sendNotification];
+}
+
+- (void)sendNotification {
+    [[NSNotificationCenter defaultCenter] postNotificationName:@"playerStateChangeNotification" object:self];
+}
+
+- (id<SPTAppRemoteTrack>) getCurrentTrackInfo {
+    return self.currentTrack;
+}
+
+- (void)startTrack {
+    [[self.appRemote playerAPI] resume:^(id result, NSError * error){
+        NSLog(@"Playing current track...");
+        if (error != nil) {
+            NSLog(@"Error: %@", error.localizedDescription);
+        } else {
+            NSLog(@"Played music");
+        }
+    }];
+}
+
+- (void)stopTrack {
+    NSLog(@"%@", @([self.appRemote isConnected]));
+    [[self.appRemote playerAPI] pause:^(id result, NSError * error){
+        NSLog(@"Pausing current track...");
+        if (error != nil) {
+            NSLog(@"Error: %@", error.localizedDescription);
+        } else {
+            NSLog(@"Stopped music.");
+        }
+    }];
+}
+
+- (void)skipTrack {
+    [[self.appRemote playerAPI] skipToNext:^(id result, NSError * error){
+        NSLog(@"Skipping current track...");
+        if (error != nil) {
+            NSLog(@"Error: %@", error.localizedDescription);
+        } else {
+            NSLog(@"Skipped to next track");
+        }
+    }];
+    
+    [self.appRemote.playerAPI skipToNext:(nil)];
 }
 
-// TODO: Add Play/Pause
-// TODO: Add Skip/Rewind
+- (void)rewindTrack {
+    NSLog(@"Rewinding music called");
+    
+    self.timestamp = 3; // Placeholder timestamp
+    
+    if (self.timestamp < MAX_MILISECONDS) { // if current timestamp < x seconds, restart current song
+        [[self.appRemote playerAPI] seekToPosition:0 callback:^(id result, NSError * error){
+            if (error != nil) {
+                NSLog(@"Error: %@", error.localizedDescription);
+            } else {
+                NSLog(@"Restarted track");
+            }
+        }];
+        
+    } else if (self.timestamp > MAX_MILISECONDS) { // if current timestamp > x seconds, rewind to previous song
+        [[self.appRemote playerAPI] skipToPrevious:^(id result, NSError * error){
+            if (error != nil) {
+                NSLog(@"Error: %@", error.localizedDescription);
+            } else {
+                NSLog(@"Skipped to previous track");
+            }
+        }];
+    }
+}
+
+- (void)retriveDataFrom:(NSString *)targetUrl result:(void (^)(NSDictionary *))parsingFinished {
+    NSString *tokenType = @"Bearer";
+    NSString *header = [NSString stringWithFormat:@"%@ %@", tokenType, self.accessToken];
+
+    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
+    
+    [request setValue:header forHTTPHeaderField:@"Authorization"];
+    [request setHTTPMethod:@"GET"];
+    [request setURL:[NSURL URLWithString:targetUrl]];
+
+    __block NSDictionary *dataRecieved = [[NSDictionary alloc] init];
```

Comment:
You don't need to declare this here. Instead, you can declare it down below on line 206 like this:

`NSDictionary *dataReceived = [NSJSONSerialization JSONObjectWithData:dataUTF8 options:0 error: &error];`